### PR TITLE
MueLu: Add cut-based dropping

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/muelu 

## Motivation
Add a cut-based dropping strategy implemented by @dsunder.

THIS IS EXPERIMENTAL!

To try it out:
```
<Parameter name="aggregation: drop scheme"             type="string" value="distance laplacian"/>
<Parameter name="aggregation: distance laplacian algo" type="string" value="scaled cut"/>
<Parameter name="aggregation: drop tol"                type="double" value="8.0"/>
```

### CODE REVIEW

File: packages/muelu/doc/UsersGuide/masterList.xml [hunk 1]
```diff
@@ -386,6 +386,15 @@
       <comment-ML>parameter not existing in ML</comment-ML>
     </parameter>
 
+    <parameter>
+      <name>aggregation: distance laplacian algo</name>
+      <type>string</type>
+      <default>"default"</default>
+      <description>Type of sub-algorithm for "distance laplacian" dropping. Possible values: "default", "unscaled cut", "scaled cut".</description>
+      <visible>true</visible>
```
Comment (jhux2) [inline packages/muelu/doc/UsersGuide/masterList.xml]: Make the visibility `false` so that it doesn't show up in the manual.

Comment (cgcgcg) [reply packages/muelu/doc/UsersGuide/masterList.xml]: That was the intent :-/
Good catch!

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_def.hpp [hunk 2]
```diff
@@ -713,28 +766,158 @@ namespace MueLu {
             numTotal += indices.size();
 
             LO nnz = indices.size(), rownnz = 0;
+
+            LO rowNonUniqueMap = nonUniqueMap->getLocalElement(uniqueMap->getGlobalElement(row));
```
[the remark this answers is not in the document]
Comment (jhux2) [reply packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_def.hpp]: As @cgcgcg I discussed earlier this week, local to global lookup is O(1), global to local requires a hash map lookup.

**Comment by jhux2** [comment]
Any update on this?

**Comment by cgcgcg** [comment]
@jhux2 I'm in the process of debugging with EMPIRE to see why rowmap != colmap on the local part.

**Comment by jhux2** [comment]
> I'm in the process of debugging with EMPIRE to see why rowmap != colmap on the local part.

Thanks!  @bathmatt was asking.

**Comment by cgcgcg** [comment]
@jhux2 This should be good to go now.

**Pull request merged**